java/src/main/native/src/RmmJni.cpp

```diff
@@ -26,12 +26,120 @@ using rmm::mr::device_memory_resource;
 
+/**
+ * @brief An RMM device memory resource that delegates to another resource
+ * while tracking the amount of memory allocated.
+ *
+ * @tparam Upstream Type of memory resource that will be wrapped.
+ * @tparam size_align The size to which all allocation requests are
+ * aligned. Must be a value >= 1.
+ */
+template <typename Upstream>
+class tracking_resource_adaptor final : public device_memory_resource {
+public:
+  /**
+   * @brief Constructs a new tracking resource adaptor that delegates to
+   * `mr` for all allocation operations while tracking the amount of memory
+   * allocated.
+   *
+   * @param mr The resource to use for memory allocation operations.
+   * @param size_alignment The alignment to which the `mr` resource will
+   * round up all memory allocation size requests.
+   */
+  tracking_resource_adaptor(Upstream* mr, std::size_t size_alignment)
+      : resource{mr}, size_align{size_alignment} {}
+
+  Upstream* get_wrapped_resource() {
+    return resource;
+  }
+
+  std::size_t get_total_allocated() {
+    std::lock_guard<std::mutex> lock(size_map_mutex);
+    return total_allocated;
+  }
+
+private:
+  Upstream* const resource;
+  std::size_t const size_align;
+  std::size_t total_allocated{0};
+
+  // map and associated lock to track memory sizes by address
+  // TODO: This should be removed when rmm::alloc and rmm::free are removed and the size parameter
+  //       for do_deallocate can be trusted. If map and mutex are removed then total_allocated
+  //       should be updated to be atomic.
+  std::unordered_map<void*, std::size_t> size_map{};
+  std::mutex size_map_mutex{};
+
+
+  void* do_allocate(std::size_t num_bytes, cudaStream_t stream) override {
+    // adjust size of allocation based on specified size alignment
+    num_bytes = (num_bytes + size_align - 1) / size_align * size_align;
+
+    std::lock_guard<std::mutex> lock(size_map_mutex);
+
+    auto result = resource->allocate(num_bytes, stream);
+    if (result) {
+      total_allocated += num_bytes;
+      size_map[result] = num_bytes;
+    }
+    return result;
+  }
+
+  void do_deallocate(void* p, std::size_t size, cudaStream_t stream) override {
+    std::lock_guard<std::mutex> lock(size_map_mutex);
+
+    resource->deallocate(p, size, stream);
+
+    if (p) {
+      // TODO: size can't be trusted until rmm::alloc and rmm::free are removed,
+      //       see https://github.com/rapidsai/rmm/issues/302
+      auto it = size_map.find(p);
+      if (it != size_map.end()) {
+        total_allocated -= it->second;
+        size_map.erase(it);
+      } else {
+        // Untracked size, may be an allocation from before resource was installed.
+      }
+    }
+  }
+
+  bool supports_get_mem_info() const noexcept override {
+    return resource->supports_get_mem_info();
+  }
+
+  std::pair<size_t, size_t> do_get_mem_info(cudaStream_t stream) const override {
+    return resource->get_mem_info(stream);
```

Comment:
Nitpick: If `tracking_resource_adaptor<Upstream>` were a derived class of `Upstream`, might we have saved having to implement `supports_get_mem_info()`, `do_get_mem_info()`, `supports_streams()` to explicitly forward to `Upstream`?

```c++
template <typename Upstream, std::enable_if<std::is_base_of<device_memory_resource, Upstream>::value>::type* = nullptr>
class tracking_resource_adaptor final : public Upstream
{...};
```

Was this disregarded because the constructor might pose a problem?

Reply:
This is a wrapping adaptor.  This allows the adaptor to be applied to an already existing instance, whereas using inheritance would not.  The code currently is wrapping a device memory resource of unknown type, so not sure how we could use inheritance here to preserve the behavior of the unknown memory resource (and its state when we replaced it).

The only reason to use a template argument here would be for the case where the code using it happens to know the specific type of device_memory_resource being wrapped so the compiler can potentially optimize what would otherwise be a virtual method invocation.  Given we are moving to a setup where the JNI layer will setup the memory resources explicitly, that's likely to be the case so I made it a template argument. This also follows the example used in the RMM memory resource adaptors, see the various adaptors at https://github.com/rapidsai/rmm/tree/branch-0.14/include/rmm/mr/device

> Was this disregarded because the constructor might pose a problem?

That's one reason.  The JNI layer needs to support both pooled mode and non-pooled mode.  A wrapping adaptor approach makes the code more reusable and easier to wield than inheritance, where we'd have to update the adaptor template each time we moved to another underlying memory resource with different construction requirements.

Reply:
> This allows the adaptor to be applied to an already existing instance, whereas using inheritance would not... A wrapping adaptor approach makes the code more reusable and easier to wield than inheritance

Good points, both.

Reply:
> This also follows the example used in the RMM memory resource adaptors, see the various adaptors at https://github.com/rapidsai/rmm/tree/branch-0.14/include/rmm/mr/device

That is informative, thanks. Keeping it in line with those other adaptors makes sense.